### PR TITLE
redis: add hashtagging

### PR DESCRIPTION
**Pull request by maximebedard**
*Description*:

Add hashtagging support to the redis-proxy filter via an opt-in setting in the connection pool options to preserve backward compatibility. The hashtagging algorithm is inspired from the [redis-cluster actual implementation](https://redis.io/topics/cluster-spec#keys-hash-tags), however instead of using a `crc16(key) % 16384`, I reused the original hashing algorithm that was being used to compute the hash key. This is mostly to ensure keys with a specific hashtag will get routed to the same upstream.

Based on the already existing code, I have a small question regarding this [comment] (https://github.com/maximebedard/envoy/blob/master/source/extensions/filters/network/redis_proxy/conn_pool_impl.h#L176). I assume this PR would involve a similar strategy? Is adding a configuration option to the public API a correct way of achieving this? I'm also not too sure if I added everything in the right places, but I'm happy to move things around to the proper classes where necessary.

Still very new to the project and to c++ in general, not entirely sure if this is the correct approach, but any guidance and feedback is truly appreciated!

*Risk Level*: Low
*Testing*: Added a unit test that verifies if the connection pool forwards the key using the expected hash key.
*Docs Changes*: Add a `hashtags` setting by connection pool to preserve backward compatibly.
*Release Notes*: N/A yet

### CODE REVIEW

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 1]
```diff
@@ -32,6 +33,9 @@ message RedisProxy {
     // is ready.
     google.protobuf.Duration op_timeout = 1
         [(validate.rules).duration.required = true, (gogoproto.stdduration) = true];
+
+    // Specifies if the connection pools must compute hashes based on hashtags
```
Comment (mattklein123) [inline api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: Sorry I don't quite understand what this does. Can you flesh out the docs?

File: source/common/config/filter_json.cc [hunk 2]
```diff
@@ -236,10 +236,10 @@ void FilterJson::translateRedisProxy(
   proto_config.set_cluster(json_config.getString("cluster_name"));
 
   const auto json_conn_pool = json_config.getObject("conn_pool");
-  json_conn_pool->validateSchema(Json::Schema::REDIS_CONN_POOL_SCHEMA);
```
Comment (mattklein123) [inline source/common/config/filter_json.cc]: v1 config is being removed soon, so please revert the changes to this file and config_schemas.cc

**Comment by maximebedard** [comment]
Thanks for the quick feedback! I've updated the documentation and reverted the deprecated API changes, let me know if this clarifies the behaviour.

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 3]
```diff
@@ -32,6 +33,18 @@ message RedisProxy {
...
+    // Use hash tagging on every redis key to guarantee that keys with the same hash tag will be
+    // forwarded to the same upstream. The hash key used for determining the upstream in a
+    // consistent hash ring configuration will be computed from the hash tagged key instead of the
+    // whole key. The algorithm used to compute the hash tag is identical to the `redis-cluster
+    // implementation <https://redis.io/topics/cluster-spec#keys-hash-tags>`_.
+    //
+    // Examples:
+    //
+    // * '{user1000}.following' and '{user1000}.followers' **will** be sent to the same upstream
+    // * '{user1000}.following' and '{user1001}.follwing' **might** be sent to the same upstream
```
Comment (mattklein123) [inline api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: typo `follwing` (I think)

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 4]
```diff
@@ -32,6 +33,18 @@ message RedisProxy {
...
+    google.protobuf.BoolValue hashtags = 2;
```
Comment (mattklein123) [inline api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: I would use a `bool` here so that it just defaults to false.

Please also add a release note referring to this feature.

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 5]
```diff
@@ -15,7 +15,32 @@ namespace ConnPool {
 
 ConfigImpl::ConfigImpl(
     const envoy::config::filter::network::redis_proxy::v2::RedisProxy::ConnPoolSettings& config)
-    : op_timeout_(PROTOBUF_GET_MS_REQUIRED(config, op_timeout)) {}
+    : op_timeout_(PROTOBUF_GET_MS_REQUIRED(config, op_timeout)),
+      hashtagger_(makeHashtagger(config.hashtags())) {}
+
+// Inspired by the redis-cluster hashtagging algorithm
+// https://redis.io/topics/cluster-spec#keys-hash-tags
+std::string ConfigImpl::hashtag(const std::string& s) {
```
Comment (danielhochman) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: does it make sense to have this as a member of ConfigImpl? i think it would be better to include it on the LbContextImpl.

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 6]
```diff
@@ -32,6 +33,18 @@ message RedisProxy {
...
+    // forwarded to the same upstream. The hash key used for determining the upstream in a
+    // consistent hash ring configuration will be computed from the hash tagged key instead of the
+    // whole key. The algorithm used to compute the hash tag is identical to the `redis-cluster
+    // implementation <https://redis.io/topics/cluster-spec#keys-hash-tags>`_.
+    //
+    // Examples:
+    //
+    // * '{user1000}.following' and '{user1000}.followers' **will** be sent to the same upstream
+    // * '{user1000}.following' and '{user1001}.following' **might** be sent to the same upstream
+    bool hashtags = 2;
```
Comment (danielhochman) [inline api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: nit: would prefer a key with a verb like `enable_hashtagging`

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc (hunk 5 above)
Comment (maximebedard) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: I was considering this too, WDYT of moving the hashtagging function into LbContextImpl, then invoking it lazily?

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 7]
```diff
@@ -308,6 +309,26 @@ void InstanceImpl::ThreadLocalActiveClient::onEvent(Network::ConnectionEvent eve
   }
 }
 
+// Inspired by the redis-cluster hashtagging algorithm
+// https://redis.io/topics/cluster-spec#keys-hash-tags
+std::string InstanceImpl::LbContextImpl::hashtag(const std::string& s, bool enabled) {
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: Can you switch this to take an `absl::string_view` and return an `absl::string_view`? If we do that we can avoid any allocations I think?

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.h [hunk 8]
```diff
@@ -172,11 +174,14 @@ class InstanceImpl : public Instance {
   };
 
   struct LbContextImpl : public Upstream::LoadBalancerContextBase {
-    LbContextImpl(const std::string& hash_key) : hash_key_(std::hash<std::string>()(hash_key)) {}
+    LbContextImpl(const std::string& key, bool enabled_hashtagging)
+        : hash_key_(std::hash<std::string>()(hashtag(key, enabled_hashtagging))) {}
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.h]: This confused me when I first saw this, so I looked at the code (the use of `std::hash` here). @HenryYYang I think this needs to get fixed also to use a configuration setting for the hash type (similar to your previous PR) or its not going to work out well when you migrate? I think this was effectively missed in the other PR. This is unrelated to the intent of this PR so no need to change here but I wanted to flag it for when we try to get rid of `std::hash` in the hashing path.

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc (hunk 7 above)
Comment (maximebedard) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: Tried implementing it using `absl::string_view`, however seems like std::string_view does not meet the requirements required for std::hash to work properly. Seems like the default constructor is deleted. :( 

```
bazel-out/darwin-fastbuild/bin/source/extensions/filters/network/redis_proxy/_virtual_includes/conn_pool_lib/extensions/filters/network/redis_proxy/conn_pool_impl.h:180:21: error: call to implicitly-deleted default constructor of 'std::hash<absl::string_view>'
        : hash_key_(std::hash<absl::string_view>()(hashtag(key, enabled_hashtagging))) {}
                    ^
```

Comment (mattklein123) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: I would go ahead and construct an `std::string` just for computing the hash, and this can be fixed when we remove `std::hash` per the TODO in the next few lines.

Comment (maximebedard) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: Sounds good 👍 I can work on that PR too, unless @HenryYYang wants to do it!

Comment (maximebedard) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: Started the WIP PR [here](https://github.com/envoyproxy/envoy/pull/5766), will rebase everything here and add the missing release note.

Comment (mattklein123) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: OK, well why don't we land this first? So update to string_view and then we can look at the other PR?

**Comment by mattklein123** [comment]
@maximebedard I think the 2 PRs are now mixed together from my look. Can we untangle them and check CI? I'm fine merging either one first, so just let me know which to look at. Thank you!

/wait

**Comment by mattklein123** [comment]
@maximebedard can you take a look at CI? I think the release failure is legit and some hash mismatch.
